src/services/unlockCourtCase.ts

```diff
@@ -6,27 +6,39 @@ import User from "./entities/User"
 const unlockCourtCase = async (
   dataSource: DataSource | EntityManager,
   courtCaseId: number,
-  user: User
+  user: User,
+  unlockReason?: "Trigger" | "Exception"
 ): Promise<UpdateResult | Error> => {
+  const { canLockExceptions, canLockTriggers, isSupervisor, username } = user
+  const shouldUnlockExceptions = canLockExceptions && (unlockReason === undefined || unlockReason === "Exception")
+  const shouldUnlockTriggers = canLockTriggers && (unlockReason === undefined || unlockReason === "Trigger")
+
+  if (!shouldUnlockExceptions && !shouldUnlockTriggers) {
+    return new Error("User hasn't got permission to unlock the case")
+  }
+
   const courtCaseRepository = dataSource.getRepository(CourtCase)
   const setFields: QueryDeepPartialEntity<CourtCase> = {}
-  const { canLockExceptions, canLockTriggers, username } = user
-
-  const query = courtCaseRepository
-    .createQueryBuilder()
-    .update(CourtCase)
-    .where({ errorLockedByUsername: username, triggerLockedByUsername: username })
+  const query = courtCaseRepository.createQueryBuilder().update(CourtCase)
 
-  if (canLockExceptions) {
+  if (shouldUnlockExceptions) {
     setFields.errorLockedByUsername = null
   }
-  if (canLockTriggers) {
+  if (shouldUnlockTriggers) {
     setFields.triggerLockedByUsername = null
   }
 
   query.set(setFields).where("error_id = :id", { id: courtCaseId })
 
-  return query.execute().catch((error: Error) => error)
+  if (!isSupervisor && shouldUnlockExceptions) {
```

Comment:
Currently, supervisors can unlock any court case given in the query parameters to the case list page. I wonder if we should prevent supervisors from unlocking cases from other forces?

Reply:
I think its low risk since they wouldn't see those cases anyway, but I refactored to make sure its absolutely impossible:
[6e94a3e](https://github.com/ministryofjustice/bichard7-next-ui/pull/165/commits/6e94a3e2261e5b17ab8a90793df77c14ce86b798)